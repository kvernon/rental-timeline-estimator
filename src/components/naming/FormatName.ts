/**
 * used for prefixing names
 * @param value
 * @param suffix
 */
export const formatName = (value: string, suffix: string): string => {
  const dot = '.';

  if (!value || value.indexOf(dot) === -1) return suffix + value;

  const strings = value.split(dot);

  strings[1] = strings[1] + dot + suffix;

  return strings.join(dot);
};
