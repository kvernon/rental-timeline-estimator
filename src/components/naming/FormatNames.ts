export const FormatNames = {
  ValidatorStackId: 'validatorStack',
  TitleDropDownValidatorId: 'titleDropDown',
  PropertyDropDownValidatorId: 'titleDropDown',
  RangeFieldValidatorId: 'rangeFieldValidator',
};
