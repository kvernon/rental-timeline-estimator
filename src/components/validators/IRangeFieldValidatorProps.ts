import { IRangeFieldValidatorEvent } from './IRangeFieldValidatorEvent';
import { FontGroups } from '../../theming/fontGroups';
import { IEventResult, IEventValue } from './IEventResult';

export interface IRangeFieldValidatorProps extends IRangeFieldValidatorEvent<IEventResult<number>> {
  direction?: 'row' | 'column';
  hasSpinner?: boolean;
  inputFontGroup?: FontGroups;
  inputLabelFontGroup?: FontGroups;
  onChange?: (inputData: IRangeFieldValidatorEvent<IEventValue<number>>) => void;
  prefix?: string;
  showTitle?: boolean;
  suffix?: string;
  title: string;
  useTransparent?: boolean;
  useUnderlineOnly?: boolean;
}
