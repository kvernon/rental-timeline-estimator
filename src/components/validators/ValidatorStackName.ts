export const ValidatorStackName = (id: string) => `validatorStack${id}`;
