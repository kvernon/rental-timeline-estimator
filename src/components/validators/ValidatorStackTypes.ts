export enum ValidatorStackTypes {
  Optional,
  Required,
}
