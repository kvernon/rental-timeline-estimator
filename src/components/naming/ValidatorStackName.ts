import { formatName } from './FormatName';
import { FormatNames } from './FormatNames';

export const ValidatorStackName = (id: string) => formatName(id, FormatNames.ValidatorStackId);
