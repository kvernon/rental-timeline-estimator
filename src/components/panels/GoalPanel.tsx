import React, { useEffect, useState } from 'react';

import { RangeFieldValidator } from '../validators/RangeFieldValidator';
import { IRangeFieldValidatorProps } from '../validators/IRangeFieldValidatorProps';
import styled from '@emotion/styled';
import { Spinner } from '../core/Spinner';
import { Stack } from '../core/Stack';

const TopStack = styled.div`
  height: 275px;
  display: block;
  width: 100%;
  overflow: visible;
  float: left;
  position: relative;
`;

const StackPosition = styled(Stack)`
  position: relative;
  width: auto;
  min-height: 0;
  z-index: 2;
`;

export const GoalPanel = <Required extends boolean = false>(props: IRangeFieldValidatorProps<Required>) => {
  const [x, setX] = useState<number | undefined>();
  const [y, setY] = useState<number | undefined>();
  const [width, setWidth] = useState<number | undefined>();
  const [height, setHeight] = useState<number | undefined>();

  const ZRangeFieldValidator = styled(RangeFieldValidator<Required>)`
    z-index: 4;
  `;

  useEffect(() => {
    const getPosition = () => {
      const ele = document.getElementById(`${props.id}-box`);

      const x = ele?.getBoundingClientRect().x;
      setX(() => x);

      const y = ele?.getBoundingClientRect().y;
      setY(() => y);

      const width = ele?.getBoundingClientRect()?.width;
      setWidth(() => width);

      const height = ele?.getBoundingClientRect()?.height;
      setHeight(() => height);
    };

    getPosition();
    window.addEventListener('resize', getPosition);

    return () => {
      window.removeEventListener('resize', getPosition);
    };
  }, [props.id]);

  return (
    <TopStack aria-label={'Goal Panel'}>
      <Spinner shape={{ x, width, y, height }} />
      <StackPosition spacing={2} paddingLeft={'20%'} paddingTop={'25px'} paddingBottom={'25px'} paddingRight={'20%'}>
        <ZRangeFieldValidator
          {...props}
          id={props.id}
          direction={'row'}
          useUnderlineOnly
          hasSpinner={false}
          useTransparent={true}
          required={props.required}
        />
      </StackPosition>
    </TopStack>
  );
};
