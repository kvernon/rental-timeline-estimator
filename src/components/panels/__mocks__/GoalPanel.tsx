import React from 'react';

import { IRangeFieldValidatorProps } from '../../validators/IRangeFieldValidatorProps';
import { ValidatorTypes } from '../../validators/ValidatorTypes';

export const GoalPanel = jest.fn((p: IRangeFieldValidatorProps<true>) => {
  return (
    <input
      type="number"
      id={p.id}
      value={p.value?.value}
      name={p.title}
      aria-label={'Goal Panel'}
      onChange={(e) => {
        if (p.onChange)
          p.onChange({
            value: parseInt(e.target.value),
            validationResult: ValidatorTypes.Valid,
          });
      }}
    />
  );
});
