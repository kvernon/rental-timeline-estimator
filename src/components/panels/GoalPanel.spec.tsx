import { configure, render, screen } from '@testing-library/react';
import { GoalPanel } from './GoalPanel';
import React from 'react';
import { IRangeFieldValidatorProps } from '../validators/IRangeFieldValidatorProps';
import { ValidatorStackTypes } from '../validators/ValidatorStackTypes';
import { ValidatorStackName } from '../naming/ValidatorStackName';

jest.mock('../validators/RangeFieldValidator', () => {
  return {
    RangeFieldValidator: jest.fn((p: IRangeFieldValidatorProps) => <input type="number" id={p.id} value={p.defaultValue} name={p.title}></input>),
  };
});

describe('GoalPanel unit test', () => {
  afterEach(() => {
    jest.resetModules();
  });

  beforeEach(() => {
    configure({ testIdAttribute: 'id' });
  });

  test('should render', () => {
    const p: IRangeFieldValidatorProps = {
      id: 'goal',
      validationType: ValidatorStackTypes.Required,
    };

    render(<GoalPanel {...p} />);

    const entity = screen.getByTestId<HTMLDivElement>(ValidatorStackName(p.id));

    expect(entity).toMatchSnapshot();
  });
});
