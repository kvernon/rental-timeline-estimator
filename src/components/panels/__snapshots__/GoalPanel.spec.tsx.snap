// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GoalPanel unit test should render 1`] = `
.emotion-0 {
  height: 275px;
  display: block;
  width: 100%;
  overflow: visible;
  float: left;
  position: relative;
}

.emotion-1 {
  width: 950px;
  height: 950px;
  left: 240px;
  top: -332px;
  float: left;
  position: absolute;
  background-image: radial-gradient(rgba(199, 159, 0, 0.3) 0%, rgba(61, 61, 61, 0.01) 62%, transparent 67%);
  border-radius: 50%;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  z-index: -1;
  -webkit-animation-name: spin;
  animation-name: spin;
  -webkit-animation-duration: 300s;
  animation-duration: 300s;
  -webkit-animation-iteration-count: infinite;
  animation-iteration-count: infinite;
  -webkit-animation-timing-function: linear;
  animation-timing-function: linear;
}

.emotion-1::after,
.emotion-1::before {
  content: '';
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  transform-origin: center;
  border-radius: 50%;
  -webkit-mask-image: radial-gradient(rgba(0, 0, 0, 1) 65%, transparent 80%);
  mask-image: radial-gradient(rgba(0, 0, 0, 1) 65%, transparent 80%);
}

.emotion-1::before {
  background: repeating-conic-gradient(from 0deg, rgb(199, 159, 0, 0.3) 0deg 10deg, transparent 10deg 15deg);
  -webkit-mask-image: radial-gradient(rgba(0, 0, 0, 0.7) 0%, rgba(0, 0, 0, 0.3) 36%, transparent 65%);
  mask-image: radial-gradient(rgba(0, 0, 0, 0.7) 0%, rgba(0, 0, 0, 0.3) 36%, transparent 65%);
  -webkit-animation: rotate 720s linear,scale 3s linear infinite;
  animation: rotate 720s linear,scale 3s linear infinite;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-flex: 2;
  -ms-flex: 2;
  flex: 2;
  padding-bottom: 25px;
  margin-bottom: 0;
  padding-top: 25px;
  padding-left: 20%;
  padding-right: 20%;
  position: relative;
  width: auto;
  min-height: 0;
  z-index: 2;
}

.emotion-3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  color: #ffffffff;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  margin: 5px 5px 24px 5px;
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: flex-start;
  justify-content: flex-start;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
}

.emotion-3:last-child {
  margin: 5px;
}

.emotion-4 {
  font-family: child;
  font-size: child;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  color: child;
  white-space: nowrap;
  padding: 5px;
  -webkit-transform: none;
  -moz-transform: none;
  -ms-transform: none;
  transform: none;
  overflow: visible;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  -webkit-box-flex: 2;
  -webkit-flex-grow: 2;
  -ms-flex-positive: 2;
  flex-grow: 2;
}

.emotion-4:focus {
  color: #9ee5ff;
}

.emotion-5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-content: space-evenly;
  -ms-flex-line-pack: space-evenly;
  align-content: space-evenly;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  height: 43px;
}

.emotion-6 {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  text-align: right;
  width: 100%;
  height: 59px;
  border-radius: 0.3rem;
  -webkit-transition: border-color 0.4s ease-out;
  transition: border-color 0.4s ease-out;
  border-color: white;
  border-width: 0px 0px 5px 0;
  font-family: child;
  font-size: child;
  color: child;
  background-color: transparent;
  border-color: #00FF00;
}

.emotion-6[type=number]::-webkit-inner-spin-button,
.emotion-6[type=number]::-webkit-outer-spin-button {
  -webkit-appearance: none;
  margin: 0;
}

.emotion-6:hover {
  border-color: #00FF00;
}

<div
  aria-label="Goal Panel"
  class="emotion-0"
>
  <div
    class="emotion-1"
    shape="[object Object]"
  />
  <div
    class="emotion-2"
    spacing="2"
  >
    <div
      aria-controls="goal"
      class="emotion-3"
      id="goal-form-control"
      role="input-control"
    >
      <label
        class="emotion-4"
        direction="row"
        for="goal"
      >
        Goal
      </label>
      <div
        class="emotion-5"
        id="goal-box"
      >
        <input
          alt="Range between 0 and 100"
          aria-description="Range between 0 and 100"
          aria-label="Goal"
          class="emotion-6"
          id="goal"
          required=""
          title="Range between 0 and 100"
          type="number"
          value="10"
        />
      </div>
    </div>
  </div>
</div>
`;
